### DISCUSSION TITLE
Integration with direnv

### DISCUSSION THREAD

**Discussion by Djaygo** [AndydeCleyre/zpy#41] [Ideas]
Hi Andyde,
Hope you are doing good! Recently I learned about `layout python` from [direnv](https://direnv.net/).
I tried to figure out how to combine it with zpy, but using for example the `eval a8 --py current` would not pick up the direnv venv. Do you know if this would be possible?

**Comment by AndydeCleyre** [comment]
Hey, thanks for asking about this. It's just Andy, btw, but I see how the username makes that unclear.

I haven't used direnv but will give it a shot to try this out. 

I *have* used [zsh-autoenv](https://github.com/Tarrasch/zsh-autoenv), which is much more straightforward for this case, as it runs any Zsh you want in the current shell, whereas direnv runs Bash and is mostly for exporting variables. 

---

So if you're using zsh-autoenv, you can have

`.autoenv.zsh`:

```zsh
a8 ${0:h}
```

`.autoenv_leave.zsh`:

```zsh
if [[ $VIRTUAL_ENV ]]  envout
```

The zero in `${0:h}` is the path of the `.autoenv.zsh` file, and the `:h` expansion gets that path's parent. This ensures the proper project folder is used, if you're activating the script by entering a deeper subdirectory.

---

Now looking at direnv... [layout python](https://github.com/direnv/direnv/blob/master/man/direnv-stdlib.1.md#layout-python-python_exe) has direnv create a new venv, which is probably not what we want (at least not where and how it does so). I guess to do things the direnv way, we want to create a new `layout_xxxx` function in `~/.config/direnv/direnvrc`.

So I'm looking at [this example](https://github.com/direnv/direnv/wiki/Python#uv) in the docs for a custom `uv`-based layout function:

```bash
layout_uv() {
    if [[ -d ".venv" ]]; then
        VIRTUAL_ENV="$(pwd)/.venv"
    fi

    if [[ -z $VIRTUAL_ENV || ! -d $VIRTUAL_ENV ]]; then
        log_status "No virtual environment exists. Executing \`uv venv\` to create one."
        uv venv
        VIRTUAL_ENV="$(pwd)/.venv"
    fi

    PATH_add "$VIRTUAL_ENV/bin"
    export UV_ACTIVE=1  # or VENV_ACTIVE=1
    export VIRTUAL_ENV
}
```

There are a few ways we could go about this, either recreating logic from zpy, or invoking Zsh+zpy from Bash. I'd lean toward the latter approach for now. Ultimately I think there are these concrete goals for the function:

- create the venv if it doesn't exist (`a8`)
- append to the `PATH`
- set and export `VIRTUAL_ENV`
- set and export `VENV_ACTIVE` (never heard that one, I guess it's a direnv thing?)

But I don't know how this is all undone cleanly, as it's not using venv's usual `activate` and `deactivate` scripts, which also alter the prompt for you.

---

Let's see what happens anyway. zpy has a helpful subcommand `mkbin` which copies the whole of zpy into a script file which runs a single function of your choice, and that's one way to handle the invoke-zpy-funcs-from-bash task. So let's do that for `a8` and `venvs_path`:

```console
$ zpy mkbin --help
Make a standalone script for any zpy function.
zpy mkbin <func> <dest>
$ zpy mkbin a8 ~/.local/bin/
$ zpy mkbin venvs_path ~/.local/bin/
```

Now I'll write a Bash function based on the above `uv`-based one, in `~/.config/direnv/direnvrc`:

```bash
layout_zpy () {
  a8
  VIRTUAL_ENV="$(venvs_path)/venv"
  PATH_add "$VIRTUAL_ENV/bin"
  export VENV_ACTIVE=1
  export VIRTUAL_ENV
}
```

OK so now I'll try adding a new `.envrc` file to a project and see what happens . . .

```console
$ print 'layout zpy' >.envrc
direnv: error /home/andy/Code/nestedtextto/.envrc is blocked. Run `direnv allow` to approve its content
$ direnv allow
direnv: loading ~/Code/nestedtextto/.envrc
direnv: export +VENV_ACTIVE +VIRTUAL_ENV ~PATH
$ which python
/home/andy/.local/share/venvs/1be7b478d580cac73a8bca0e01d57a97/venv/bin/python
$ cd ..
direnv: unloading
$ which python
/home/andy/.local/share/mise/installs/python/latest/bin/python
```

Looking good! And at least with, for example, p10k prompt for Zsh, the venv will be reflected there automatically.

**Comment by Djaygo** [comment]
Hi Andy,

Thank you for the solution, its working great! I should have included my attempt using `direnvrc`, it could have given you a headstart.

Best regards,
Djaygo

**Reply by AndydeCleyre** [reply]
Thanks for confirming -- I've added [a docs page](https://zpy.readthedocs.io/en/latest/direnv/) based on this.
